modules/javafx.graphics/src/main/java/com/sun/scenario/animation/AbstractPrimaryTimer.java

```diff
@@ -308,29 +317,80 @@ protected void timePulseImpl(long now) {
             debugNanos += fixedPulseLength;
             now = debugNanos;
         }
-        final PulseReceiver receiversSnapshot[] = receivers;
+
+        final ReceiverRecord<PulseReceiver>[] receiversSnapshot = receivers;
         final int rLength = receiversLength;
-        try {
-            receiversLocked = true;
-            for (int i = 0; i < rLength; i++) {
-                receiversSnapshot[i].timePulse(TickCalculation.fromNano(now));
+        receiversLocked = true;
+
+        for (int i = 0; i < rLength; i++) {
+            try {
+                receiversSnapshot[i].receiver().timePulse(TickCalculation.fromNano(now));
+            } catch (Throwable e) {
+                receiversSnapshot[i].handleException(e);
             }
-        } finally {
-            receiversLocked = false;
         }
+
+        receiversLocked = false;
         recordAnimationEnd();
 
-        final TimerReceiver animationTimersSnapshot[] = animationTimers;
+        final ReceiverRecord<TimerReceiver>[] animationTimersSnapshot = animationTimers;
         final int aTLength = animationTimersLength;
-        try {
-            animationTimersLocked = true;
-            // After every frame, call any frame jobs
-            for (int i = 0; i < aTLength; i++) {
-                animationTimersSnapshot[i].handle(now);
+        animationTimersLocked = true;
+
+        // After every frame, call any frame jobs
+        for (int i = 0; i < aTLength; i++) {
+            try {
+                animationTimersSnapshot[i].receiver().handle(now);
+            } catch (Throwable e) {
+                animationTimersSnapshot[i].handleException(e);
             }
-        } finally {
-            animationTimersLocked = false;
         }
+
+        animationTimersLocked = false;
     }
 
+    private static abstract class ReceiverRecord<T> {
+
+        static ReceiverRecord<TimerReceiver> ofAnimationTimer(TimerReceiver receiver) {
+            return new ReceiverRecord<>() {
+                @Override TimerReceiver receiver() { return receiver; }
+                @Override Class<?> type() { return AnimationTimer.class; }
+            };
+        }
+
+        static ReceiverRecord<PulseReceiver> ofPulseReceiver(PulseReceiver receiver) {
+            return new ReceiverRecord<>() {
+                @Override PulseReceiver receiver() { return receiver; }
+                @Override Class<?> type() { return PulseReceiver.class; }
+            };
+        }
+
+        abstract T receiver();
+        abstract Class<?> type();
+
+        int exceptionsThrown;
+
+        void handleException(Throwable e) {
+            if (exceptionsThrown < FAILING_TIMER_THRESHOLD) {
+                exceptionsThrown++;
+
+                try {
+                    Thread thread = Thread.currentThread();
+                    thread.getUncaughtExceptionHandler().uncaughtException(thread, e);
+                } catch (Throwable ignored) {
```

Comment:
any reason not to wrap the whole method in `try`/`catch`?

Reply:
There isn't anything else in there that could throw, except for the logger itself. I don't think it makes sense to try to account for failing loggers, as basically everything will break if that's the case.